File: sm64py/billboard.py
"""Turning an actor's billboarded parts to face the camera.

SM64 draws some actor parts -- a goomba's face, most of a scuttlebug's body --
as flat quads that the original rebuilds every frame to point at the camera.
glTF has no billboard concept, so they arrive as ordinary geometry, and
Panda3D's own billboard effect cannot help: it acts on a node's transform, and
this geometry is skinned to character joints. The exporter therefore makes each
such quad a joint of its own, and they get driven from here.

The part that took several wrong turns to find is the frame the driving happens
in. `Actor.control_joint` hands back a NodePath parented to the *model root*,
not into the joint hierarchy -- so its transform reads as the joint's local
value while its scene-graph parent is somewhere else entirely. Two consequences,
both of which produced plausible-looking code that did nothing:

  * `set_hpr(some_other_node, ...)` is not a way to escape the joint's parents.
    Panda3D solves that against the *scene graph* parent, which is the model
    root, so it comes out identical to the plain local call. It never sees the
    joint chain at all.

  * The joint chain's rotation is still applied, inside the Character, on top of
    whatever is set here. On the goomba that hidden rotation is about a quarter
    turn of roll, which turns a local heading into net *pitch* -- so heading
    tipped the quad up and down instead of turning it about vertical, and no
    value of it could ever have worked.

So the rotation wanted in world space is composed against the inverse of the
parent joint's measured net rotation. `net = local * parent`, hence
`local = world * parent^-1`. Measured on an isolated goomba face, that takes the
width it holds around an orbit from 0.13 of its widest to 0.84 -- the remainder
being perspective, since the quad does not sit on the axis it orbits.

Everything the aiming depends on lives in `Tuning` so it can be adjusted from
the asset workbench and written back out, rather than being guessed at in
source. See tools/workbench.py.
"""

import json
import math
import os

# name, default, step for interactive adjustment, what it does
FIELDS = (
    ("enabled", True, None,
     "drive these joints at all"),
    ("cancel_parent", True, None,
     "compose against the parent joint's inverse rotation; without this the "
     "quad turns about whatever axis the joint chain leaves it with"),
    ("heading_offset", 0.0, 5.0,
     "degrees added to the heading that points at the camera, for quads whose "
     "authored facing is not straight down -Y"),
    # Neither of these changes the silhouette of a flat quad facing the
    # camera -- pitch tips it away, roll spins it about its own normal -- so
    # both stay at zero. They are here because getting that wrong is what the
    # earlier attempts were, and because a quad that is not flat would need
    # them.
    ("pitch", 0.0, 5.0, "world pitch held while facing the camera"),
    ("roll", 0.0, 5.0, "world roll held while facing the camera"),
    ("scale", 4.0, 0.25,
     "billboarded geometry escapes the GEO_SCALE(0x00, 16384) wrapping these "
     "actors, because the original rebuilds the matrix at a billboard rather "
     "than accumulating into it; 4.0 is exactly 1/0.25 and puts it back"),
)

DEFAULTS = {name: value for name, value, _, _ in FIELDS}
STEPS = {name: step for name, _, step, _ in FIELDS}
HELP = {name: text for name, _, _, text in FIELDS}

DEFAULT_PATH = os.path.join(
    os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
    "assets", "billboard_tuning.json")


class Tuning:
    """Billboard settings, globally and per actor.

    Per-actor entries hold only the fields that differ, so a change to a
    default is picked up by every actor that has not been overridden.
    """

    def __init__(self, base=None, per_actor=None):
        self.base = dict(DEFAULTS)
        if base:
            self.base.update(base)
        self.per_actor = {k: dict(v) for k, v in (per_actor or {}).items()}

    # -- reading and writing ------------------------------------------------

    def get(self, field, actor=None):
        if actor and field in self.per_actor.get(actor, {}):
            return self.per_actor[actor][field]
        return self.base.get(field, DEFAULTS.get(field))

    def set(self, field, value, actor=None):
        if actor:
            self.per_actor.setdefault(actor, {})[field] = value
        else:
            self.base[field] = value

    def nudge(self, field, direction, actor=None):
        """Step a numeric field, or flip a boolean one."""
        current = self.get(field, actor)
        if isinstance(current, bool):
            self.set(field, not current, actor)
        else:
            step = STEPS.get(field) or 1.0
            self.set(field, round(current + step * direction, 4), actor)
        return self.get(field, actor)

    def clear(self, actor=None):
        if actor:
            self.per_actor.pop(actor, None)
        else:
            self.base = dict(DEFAULTS)

    def to_dict(self):
        return {"base": self.base, "per_actor": self.per_actor}

    def describe(self, actor=None):
        """One line per field, marking which are actor overrides."""
        lines = []
        for name, _, _, _ in FIELDS:
            override = actor and name in self.per_actor.get(actor, {})
            value = self.get(name, actor)
            shown = value if not isinstance(value, float) else f"{value:g}"
            lines.append(f"{name:<15} {str(shown):>8}"
                         f"{'  (override)' if override else ''}")
        return lines

    # -- files --------------------------------------------------------------

    @classmethod
    def load(cls, path=None):
        """Read tuning from disk, falling back to the defaults above."""
        path = path or DEFAULT_PATH
        try:
            with open(path) as handle:
                data = json.load(handle)
        except (OSError, ValueError):
            return cls()
        return cls(data.get("base"), data.get("per_actor"))

    def save(self, path=None):
        path = path or DEFAULT_PATH
        os.makedirs(os.path.dirname(path), exist_ok=True)
        with open(path, "w") as handle:
            json.dump(self.to_dict(), handle, indent=2, sort_keys=True)
            handle.write("\n")
        return path


class Rig:
    """One billboard joint, and everything needed to aim it every frame."""

    def __init__(self, name, control, parent, exposed, rest_pos,
                 actor_name=None, owner=None):
        self.name = name
        self.control = control
        # The parent joint, exposed. Read live rather than cached because an
        # animated parent moves, and the correction has to move with it.
        self.parent = parent
        # This joint, exposed. Reports where the quad ended up after the
        # character was evaluated, which is what a heading has to be measured
        # from.
        self.exposed = exposed
        # In the parent's frame, which is where a joint's local translation
        # lives. Not the difference of the two net positions: that is the same
        # vector expressed in the actor's frame, and the two only agree while
        # the parent is unrotated -- which it is not.
        self.rest_pos = rest_pos
        self.actor_name = actor_name
        self.owner = owner

    def world_pos(self, reference):
        """Where this quad actually is, for working out a heading from.

        Not the control node's position: that reads as the joint's local
        offset, which is measured from a parent that is elsewhere.
        """
        if self.exposed is not None:
            return self.exposed.get_pos(reference)
        return self.control.get_pos(reference)

    def aim(self, reference, target, tuning):
        """Point the quad at `target`, both given in `reference`'s space."""
        from panda3d.core import Mat4, Quat, VBase3

        actor = self.actor_name
        if not tuning.get("enabled", actor):
            return

        here = self.world_pos(reference)
        heading = math.degrees(math.atan2(-(target[0] - here[0]),
                                          target[1] - here[1]))
        heading += tuning.get("heading_offset", actor)

        world = Quat()
        world.set_hpr(VBase3(heading, tuning.get("pitch", actor),
                             tuning.get("roll", actor)))

        if tuning.get("cancel_parent", actor) and self.parent is not None:
            # Only the rotation, taken from hpr rather than from the matrix, so
            # the parent's baked 0.25 scale stays out of it.
            inverse = Quat()
            inverse.set_hpr(self.parent.get_hpr(reference))
            inverse.invert_in_place()
            world = world * inverse

        mat = Mat4()
        world.extract_to_matrix(mat)
        scale = tuning.get("scale", actor)
        for row in range(3):
            mat.set_row(row, mat.get_row3(row) * scale)
        mat.set_row(3, self.rest_pos)
        self.control.set_mat(mat)


def claim(actor, prefix="billboard_", actor_name=None, owner=None):
    """Take over every joint the exporter marked as a billboard.

    Returns a Rig per joint. The joint's rest translation is worked out before
    control is taken, because `control_joint` initialises its node to identity
    and would otherwise silently drop any authored offset.
    """
    from panda3d.core import Point3

    # Exposed joints report identity until the character has been evaluated
    # once, and a rest offset measured against identity is silently wrong.
    actor.update(force=True)

    parents = joint_parents(actor)
    rigs = []
    for joint in actor.get_joints():
        name = joint.get_name()
        if not name.startswith(prefix):
            continue

        parent = expose(actor, parents.get(name))

        # Rest offset, in the parent's frame, read while the joint is still
        # animated normally.
        exposed = expose(actor, name)
        rest = Point3(0, 0, 0)
        if exposed is not None and parent is not None:
            rest = Point3(exposed.get_pos(parent))

        control = actor.control_joint(None, "modelRoot", name)
        if control is None:
            continue
        rigs.append(Rig(name, control, parent, exposed, rest, actor_name,
                        owner))
    return rigs


def expose(actor, name):
    """Expose a joint, or None if that name is not one.

    The top of the hierarchy is a PartGroup rather than a CharacterJoint and
    has no net transform to report, so a joint at the root has nothing to
    cancel against.
    """
    if not name:
        return None
    try:
        return actor.expose_joint(None, "modelRoot", name)
    except AttributeError:
        return None


def joint_parents(actor):
    """Joint name -> parent joint name, for the whole character."""
    parents = {}

    def walk(node, parent):
        parents[node.get_name()] = parent
        for i in range(node.get_num_children()):
            walk(node.get_child(i), node.get_name())

    walk(actor.get_part_bundle("modelRoot"), None)
    return parents


def probe(actor, prefix="billboard_"):
    """Everything measurable about an actor's billboard joints.

    Written for reading, by a person or by an agent with no screen: if the
    aiming is wrong again, this is the evidence to reason from rather than
    another round of guessing at constants.
    """
    from panda3d.core import GeomVertexReader, Vec3

    # Same trap as in claim(), and worse here: a probe that reports every
    # parent rotation as identity is exactly the kind of confident wrong
    # measurement this whole tool exists to prevent.
    actor.update(force=True)

    parents = joint_parents(actor)
    geometry = _geometry_by_joint(actor, GeomVertexReader, Vec3)

    rows = []
    for joint in actor.get_joints():
        name = joint.get_name()
        exposed = expose(actor, name)
        parent = expose(actor, parents.get(name))
        verts = geometry.get(name, [])
        extent = [0.0, 0.0, 0.0]
        normal = None
        if verts:
            lo = [min(v[i] for v, _ in verts) for i in range(3)]
            hi = [max(v[i] for v, _ in verts) for i in range(3)]
            extent = [round(hi[i] - lo[i], 2) for i in range(3)]
            acc = Vec3(0, 0, 0)
            for _, n in verts:
                acc += n
            if acc.length() > 1e-6:
                acc.normalize()
                normal = [round(acc[i], 3) for i in range(3)]

        # A quad is flat along exactly one axis, and that axis is its normal.
        # Worth reporting separately because the stored normals are not always
        # trustworthy -- several of these assets export a uniform placeholder.
        flat = [i for i, size in enumerate(extent) if size < 0.01]
        rows.append({
            "joint": name,
            "billboard": name.startswith(prefix),
            "parent": parents.get(name),
            "vertices": len(verts),
            "extent": extent,
            "flat_axis": "xyz"[flat[0]] if len(flat) == 1 and verts else None,
            "stored_normal": normal,
            "net_pos": _round(exposed.get_pos(actor) if exposed else None),
            "net_hpr": _round(exposed.get_hpr(actor) if exposed else None),
            "net_scale": _round(exposed.get_scale(actor) if exposed else None, 3),
            "parent_hpr": _round(parent.get_hpr(actor) if parent else None),
        })
    return rows


def _round(vector, places=2):
    return [round(v, places) for v in vector] if vector is not None else None


def _geometry_by_joint(actor, GeomVertexReader, Vec3):
    """Vertices and normals grouped by the joint they are weighted to.

    SM64 actors are rigidly segmented rather than smooth-skinned, so every
    vertex belongs to exactly one joint and this grouping is exact.
    """
    per_joint = {}
    for holder in actor.find_all_matches("**/+GeomNode"):
        node = holder.node()
        for index in range(node.get_num_geoms()):
            vdata = node.get_geom(index).get_vertex_data()
            table = vdata.get_transform_blend_table()
            if table is None:
                continue
            names = []
            for blend_index in range(table.get_num_blends()):
                blend = table.get_blend(blend_index)
                names.append(blend.get_transform(0).get_joint().get_name()
                             if blend.get_num_transforms() else None)

            vertices = GeomVertexReader(vdata, "vertex")
            normals = GeomVertexReader(vdata, "normal")
            blends = GeomVertexReader(vdata, "transform_blend")
            while not vertices.is_at_end():
                position = Vec3(vertices.get_data3())
                normal = Vec3(normals.get_data3())
                name = names[blends.get_data1i()]
                if name:
                    per_joint.setdefault(name, []).append((position, normal))
    return per_joint

File: src/billboard.rs
//! Turning billboarded geometry to face the camera.
//!
//! SM64 draws some things as flat quads it rebuilds every frame to point at
//! the camera: whole objects, like the trees, and single parts of an actor,
//! like a scuttlebug's eyes. glTF has no billboard concept, so all of it arrives
//! as ordinary geometry, drawn from whatever one side it was authored on. A
//! flat quad seen from ninety degrees away is not merely wrong, it is nothing
//! at all -- the tree meshes here measure exactly zero thick.
//!
//! Ported from `sm64py/billboard.py` and the `billboard == "axis"` branch of
//! `sm64py/level.py`, which are also where the reasoning below was worked out
//! the hard way.
//!
//! Two mechanisms, because there are two cases:
//!
//!   * A **whole object** is plain geometry and turns bodily. Only about the
//!     vertical, so a tree does not tip over when the camera looks down at it.
//!   * A **part of an actor** is skinned to a joint, so no transform on the
//!     object can reach it. The exporter makes each such quad a joint of its
//!     own, named `billboard_*`, and those are driven individually.

use bevy::{
    ecs::{schedule::ScheduleConfigs, system::ScheduleSystem},
    prelude::*,
    transform::TransformSystems,
};

/// The name the exporter gives a joint that is a billboard quad.
const JOINT_PREFIX: &str = "billboard_";

/// Scale put back onto a billboard joint.
///
/// These actors are wrapped in a `GEO_SCALE(0x00, 16384)` of 0.25, which the
/// exporter bakes onto the root joint and every joint under it inherits.
/// Billboarded geometry escapes it in the original, because there the matrix
/// is rebuilt at the billboard rather than accumulated into it. 4.0 is exactly
/// 1/0.25 and puts it back.
const JOINT_SCALE: f32 = 4.0;

/// An object drawn as a flat quad that turns bodily to face the camera.
#[derive(Component)]
pub struct BillboardAxis;

/// An actor that may carry billboard joints inside its skeleton, and whose
/// surfaces are therefore drawn from both sides.
#[derive(Component)]
pub struct BillboardActor;

/// One quad inside an actor, claimed from the name its exporter gave it.
#[derive(Component)]
pub struct BillboardJoint;

/// Tags every joint the exporter marked as a billboard, as its scene arrives.
///
/// By name, because that is the only thing that survives the export -- the
/// same reason the animation clips are looked up by name.
pub fn claim(mut commands: Commands, arrivals: Query<(Entity, &Name), Added<Name>>) {
    for (entity, name) in &arrivals {
        if name.as_str().starts_with(JOINT_PREFIX) {
            commands.entity(entity).insert(BillboardJoint);
        }
    }
}

/// Draws billboarded surfaces from both sides.
///
/// A billboard quad is single-sided, and which of its faces was authored
/// toward the viewer is not something this port can see. Aimed the wrong way
/// round it would be invisible from every angle rather than from half of them,
/// which is a worse failure than the one being fixed -- and the original never
/// culls these anyway, since it never sees the back of one.
pub fn two_sided(
    mut materials: ResMut<Assets<StandardMaterial>>,
    hierarchy: Query<&ChildOf>,
    actors: Query<(), With<BillboardActor>>,
    surfaces: Query<
        (Entity, &MeshMaterial3d<StandardMaterial>),
        Added<MeshMaterial3d<StandardMaterial>>,
    >,
) {
    for (entity, handle) in &surfaces {
        let mut ancestor = entity;
        loop {
            if actors.contains(ancestor) {
                if let Some(mut material) = materials.get_mut(&handle.0) {
                    material.double_sided = true;
                    material.cull_mode = None;
                }
                break;
            }
            let Ok(parent) = hierarchy.get(ancestor) else {
                break;
            };
            ancestor = parent.parent();
        }
    }
}

/// The rotation that points a quad's authored face at a target.
///
/// About the vertical only. Pitch would tip the quad away from the camera and
/// roll would spin it about its own normal, and neither changes the
/// silhouette of something flat -- so the one useful degree of freedom is the
/// one taken here.
///
/// The quads face +Z as authored: they come from a build whose forward is
/// Panda3D's -Y, and that is +Z once the axes are converted.
pub fn facing(here: Vec3, target: Vec3) -> Quat {
    let away = target - here;
    Quat::from_rotation_y(away.x.atan2(away.z))
}

/// Aims everything billboarded at the camera, once per rendered frame.
#[allow(clippy::type_complexity)]
pub fn aim(
    // **The eye the player is actually looking from, which is not always where
    // the camera entity is.** A gate the boom goes through flies the camera to
    // the far end of the pair for the frame it is drawn on -- see
    // [`crate::portal::carry_camera`] -- and a card turned to face *that* is a
    // card turned across the map. `FollowCamera::eye` is where the view
    // logically is, beside the player, which is the right answer for the
    // overwhelming majority of what is on the screen at that moment.
    //
    // It is `Option` because the impostor baker runs this same chain with a
    // camera of its own and no follow rig on it; there the entity's own
    // transform is the only eye there is.
    camera: Query<
        (&GlobalTransform, Option<&crate::camera::FollowCamera>),
        (
            With<Camera3d>,
            Without<BillboardAxis>,
            Without<crate::portal::PortalView>,
        ),
    >,
    globals: Query<&GlobalTransform>,
    mut objects: Query<(&mut Transform, &GlobalTransform), With<BillboardAxis>>,
    mut joints: Query<
        (&mut Transform, &GlobalTransform, &ChildOf),
        (With<BillboardJoint>, Without<BillboardAxis>),
    >,
) {
    let Ok((view, follow)) = camera.single() else {
        return;
    };
    let eye = follow.map_or_else(|| view.translation(), |follow| follow.eye.translation);
    for (mut transform, global) in &mut objects {
        transform.rotation = facing(global.translation(), eye);
    }
    for (mut transform, global, parent) in &mut joints {
        let Ok(above) = globals.get(parent.parent()) else {
            continue;
        };
        // The joint chain above this quad leaves it with a rotation of its
        // own -- on the scuttlebug, about a quarter turn of roll -- and a local
        // value is applied on top of that. So a heading set here comes out as
        // net *pitch*, and no value of it could ever have worked. Since
        // `net = parent * local`, the local wanted is `parent^-1 * world`.
        let (_, above_rotation, _) = above.to_scale_rotation_translation();
        transform.rotation = above_rotation.inverse() * facing(global.translation(), eye);
        transform.scale = Vec3::splat(JOINT_SCALE);
    }
}

/// Aiming runs after the animation player has posed the skeleton and before
/// the transforms are propagated.
///
/// Both halves are load-bearing. Before the animation player, every joint
/// written here is overwritten by the clip a moment later and nothing turns at
/// all; after propagation, the turn is a frame late and, worse, the quads are
/// drawn from transforms that never saw it.
pub fn systems() -> ScheduleConfigs<ScheduleSystem> {
    (claim, two_sided, aim)
        .chain()
        .after(bevy::animation::animate_targets)
        .before(TransformSystems::Propagate)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A quad facing +Z turns to put that face on the camera, whichever side
    /// of it the camera is on.
    #[test]
    fn a_quad_turns_its_face_to_the_camera() {
        let here = Vec3::new(4.0, 1.0, -7.0);
        for camera in [
            Vec3::new(4.0, 1.0, 0.0),
            Vec3::new(-20.0, 3.0, -7.0),
            Vec3::new(30.0, 0.0, -30.0),
            Vec3::new(4.0, 40.0, -12.0),
        ] {
            let face = facing(here, camera) * Vec3::Z;
            let wanted = (camera - here) * Vec3::new(1.0, 0.0, 1.0);
            let cosine = face.normalize().dot(wanted.normalize());
            assert!(
                cosine > 0.999,
                "the quad faces {face:?} with the camera at {camera:?}"
            );
        }
    }

    /// It turns about the vertical only: a camera looking down from overhead
    /// must not tip a tree onto its back.
    #[test]
    fn a_billboard_never_tips_over() {
        let up = facing(Vec3::ZERO, Vec3::new(0.3, 50.0, 0.3)) * Vec3::Y;
        assert!(
            (up - Vec3::Y).length() < 1e-5,
            "the billboard's up ended at {up:?}"
        );
    }

    /// The joint case: whatever rotation the skeleton leaves the quad with is
    /// cancelled, so the quad ends up facing the camera in world space rather
    /// than relative to a bone that is itself rotated a quarter turn.
    #[test]
    fn a_joints_parent_rotation_is_cancelled() {
        // A quarter turn of roll, which is what a billboard quad hangs off.
        let parent = Quat::from_rotation_z(std::f32::consts::FRAC_PI_2);
        let here = Vec3::new(1.0, 2.0, 3.0);
        let camera = Vec3::new(9.0, 2.5, -4.0);
        let local = parent.inverse() * facing(here, camera);
        let net = parent * local;
        let face = net * Vec3::Z;
        let wanted = (camera - here) * Vec3::new(1.0, 0.0, 1.0);
        assert!(
            face.normalize().dot(wanted.normalize()) > 0.999,
            "the quad ended up facing {face:?} rather than the camera"
        );
    }

    /// Claiming is by name, so the name is worth pinning: if the exporter ever
    /// stops marking these, nothing turns and nothing says why.
    ///
    /// No actor the game loads carries one any more. The goomba and the
    /// scuttlebug did -- both were decomp exports, and their faces and eyes were
    /// the flat quads SM64 rebuilds every frame -- and the slime and the ant
    /// that replaced them are authored art with real skinned meshes and nothing
    /// flat anywhere on them. So the name is pinned against the decomp's own
    /// export, kept unedited under `assets/packs/reference`, and the loaded
    /// actors are checked for the opposite: one arriving with a `billboard_*`
    /// joint means this module is back in the business of driving joints, which
    /// is worth hearing from a test rather than from a face pointing the wrong
    /// way.
    #[test]
    fn the_actors_still_carry_the_joints_this_claims() {
        let nodes = node_names("packs/reference/actors/scuttlebug.glb");
        assert!(
            nodes.iter().any(|name| name.starts_with(JOINT_PREFIX)),
            "the decomp scuttlebug has no {JOINT_PREFIX}* joint for this to drive"
        );
        // The tree is the other case: a whole object, with no joint of its own
        // to claim, turned bodily instead.
        let tree = node_names("actors/tree.glb");
        assert!(
            !tree.iter().any(|name| name.starts_with(JOINT_PREFIX)),
            "the tree grew a billboard joint, so it wants driving the other way"
        );
        // And the two enemies are a third case: actors with no billboarded
        // part at all. They still carry `BillboardActor`, which is what makes
        // their surfaces double-sided, but there is nothing here for `aim` to
        // turn.
        for actor in ["slime", "ant"] {
            let nodes = node_names(&format!("actors/{actor}.glb"));
            assert!(
                !nodes.iter().any(|name| name.starts_with(JOINT_PREFIX)),
                "the {actor} grew a billboard joint, so something has to drive it"
            );
        }
    }

    /// Where `JOINT_SCALE` comes from. If the exporter ever stops baking the
    /// 0.25 onto the skeleton, putting 4.0 back would make these quads four
    /// times too big rather than the right size.
    ///
    /// Measured on the decomp export rather than on a runtime actor, for the
    /// same reason as above: nothing the game loads carries a billboard joint
    /// now. It is also the only copy that still carries the 0.25 at all -- the
    /// runtime scuttlebug was rebuilt from a .blend and came back with no scale
    /// on its skeleton, so this had been failing against it since before the
    /// bug was replaced.
    #[test]
    fn the_scale_put_back_is_the_one_baked_in() {
        let actor = "packs/reference/actors/scuttlebug";
        let baked = root_joint_scale(&format!("{actor}.glb"))
            .unwrap_or_else(|| panic!("{actor} has no scale on its skeleton"));
        assert!(
            (baked * JOINT_SCALE - 1.0).abs() < 1e-4,
            "{actor} bakes {baked} and this puts back {JOINT_SCALE}"
        );
    }

    /// A tree is a flat card in the XY plane, so its face is along Z -- which
    /// is the whole assumption `facing` turns it on, and the reason it
    /// disappears entirely when it is not turned.
    #[test]
    fn a_tree_is_a_flat_card_facing_along_z() {
        let extent = mesh_extent("actors/tree.glb");
        assert_eq!(extent[2], 0.0, "the tree is {extent:?}, not flat along Z");
        assert!(
            extent[0] > 0.0 && extent[1] > 0.0,
            "the tree has no face at all: {extent:?}"
        );
    }

    /// Aiming turns the objects in the world, not just the maths above.
    #[test]
    fn aiming_turns_a_field_of_billboards_toward_the_camera() {
        use bevy::ecs::system::RunSystemOnce;

        let mut world = World::new();
        let eye = Vec3::new(12.0, 4.0, -9.0);
        world.spawn((
            Camera3d::default(),
            Transform::from_translation(eye),
            GlobalTransform::from_translation(eye),
        ));
        let places = [
            Vec3::new(0.0, 0.0, 0.0),
            Vec3::new(-8.0, 1.0, 4.0),
            Vec3::new(20.0, 0.0, -30.0),
        ];
        let trees: Vec<_> = places
            .iter()
            .map(|place| {
                world
                    .spawn((
                        BillboardAxis,
                        Transform::from_translation(*place),
                        GlobalTransform::from_translation(*place),
                    ))
                    .id()
            })
            .collect();
        world.run_system_once(aim).expect("aim could not run");
        for (tree, place) in trees.iter().zip(places) {
            let face = world.get::<Transform>(*tree).unwrap().rotation * Vec3::Z;
            let wanted = ((eye - place) * Vec3::new(1.0, 0.0, 1.0)).normalize();
            assert!(
                face.dot(wanted) > 0.999,
                "the tree at {place:?} faces {face:?} rather than {wanted:?}"
            );
        }
    }

    // -- reading the glTF directly, so none of this needs a renderer ---------

    fn gltf(path: &str) -> serde_json::Value {
        let root = std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("assets");
        let bytes = std::fs::read(root.join(path)).expect("missing glb");
        let length = u32::from_le_bytes(bytes[12..16].try_into().unwrap()) as usize;
        serde_json::from_slice(&bytes[20..20 + length]).expect("bad glb json")
    }

    fn node_names(path: &str) -> Vec<String> {
        gltf(path)["nodes"]
            .as_array()
            .map(|nodes| {
                nodes
                    .iter()
                    .filter_map(|node| node["name"].as_str().map(str::to_owned))
                    .collect()
            })
            .unwrap_or_default()
    }

    /// The uniform scale the exporter bakes onto the skeleton, if it does.
    fn root_joint_scale(path: &str) -> Option<f32> {
        gltf(path)["nodes"]
            .as_array()?
            .iter()
            .find_map(|node| Some(node["scale"].as_array()?[0].as_f64()? as f32))
    }

    /// How big the first mesh primitive is along each axis, from the accessor
    /// bounds the exporter writes.
    fn mesh_extent(path: &str) -> [f32; 3] {
        let file = gltf(path);
        let accessor = file["meshes"][0]["primitives"][0]["attributes"]["POSITION"]
            .as_u64()
            .expect("no positions") as usize;
        let bounds = |key: &str| -> Vec<f32> {
            file["accessors"][accessor][key]
                .as_array()
                .expect("the exporter wrote no bounds")
                .iter()
                .map(|value| value.as_f64().unwrap() as f32)
                .collect()
        };
        let (low, high) = (bounds("min"), bounds("max"));
        [high[0] - low[0], high[1] - low[1], high[2] - low[2]]
    }
}
